eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs

```diff
@@ -868,26 +895,158 @@ ppDRepDepositL = ppLens . hkdDRepDepositL @era @Identity
 ppDRepActivityL = ppLens . hkdDRepActivityL @era @Identity
 
-ppuPoolVotingThresholdsL :: forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe PoolVotingThresholds)
+ppuPoolVotingThresholdsL ::
+  forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe PoolVotingThresholds)
 ppuPoolVotingThresholdsL = ppuLens . hkdPoolVotingThresholdsL @era @StrictMaybe
 
-ppuDRepVotingThresholdsL :: forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe DRepVotingThresholds)
+ppuDRepVotingThresholdsL ::
+  forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe DRepVotingThresholds)
 ppuDRepVotingThresholdsL = ppuLens . hkdDRepVotingThresholdsL @era @StrictMaybe
 
-ppuCommitteeMinSizeL :: forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe Natural)
+ppuCommitteeMinSizeL ::
+  forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe Natural)
 ppuCommitteeMinSizeL = ppuLens . hkdCommitteeMinSizeL @era @StrictMaybe
 
-ppuCommitteeMaxTermLengthL :: forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe Natural)
+ppuCommitteeMaxTermLengthL ::
+  forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe Natural)
 ppuCommitteeMaxTermLengthL = ppuLens . hkdCommitteeMaxTermLengthL @era @StrictMaybe
 
-ppuGovActionLifetimeL :: forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe EpochNo)
+ppuGovActionLifetimeL ::
+  forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe EpochNo)
 ppuGovActionLifetimeL = ppuLens . hkdGovActionLifetimeL @era @StrictMaybe
 
-ppuGovActionDepositL :: forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe Coin)
+ppuGovActionDepositL ::
+  forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe Coin)
 ppuGovActionDepositL = ppuLens . hkdGovActionDepositL @era @StrictMaybe
 
-ppuDRepDepositL :: forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe Coin)
+ppuDRepDepositL ::
+  forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe Coin)
 ppuDRepDepositL = ppuLens . hkdDRepDepositL @era @StrictMaybe
 
-ppuDRepActivityL :: forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe EpochNo)
+ppuDRepActivityL ::
+  forall era. ConwayEraPParams era => Lens' (PParamsUpdate era) (StrictMaybe EpochNo)
 ppuDRepActivityL = ppuLens . hkdDRepActivityL @era @StrictMaybe
+
+ppUpdate ::
+  PParamsUpdate era ->
+  Lens' (PParamsUpdate era) (StrictMaybe a) ->
+  Lens' (PParams era) a ->
+  PParams era ->
+  PParams era
+ppUpdate ppu ppuL ppL pp =
+  case ppu ^. ppuL of
+    SNothing -> pp
+    SJust p -> pp & ppL .~ p
+
+ppUpdateCostModels ::
+  AlonzoEraPParams era =>
+  PParamsUpdate era ->
+  PParams era ->
+  PParams era
+ppUpdateCostModels ppu pp =
+  case ppu ^. ppuCostModelsL of
+    SNothing -> pp
```

Comment:
How come this pparam gets special treatment if it's empty, as opposed to other ppu lenses of type StrictMaybe?

Reply:
Good question. Here is why: #3631 and the fix #3771

Reply:
Great, I see now the connection, thanks :+1: